File: python/OECD-Common/OecdDataResponse.py
import datetime
import uuid
from typing import List, Dict, Any


# https://stackoverflow.com/questions/53905951/how-do-i-serialise-a-nested-dictionary-with-marshmallow
class OecdDataResponseValue:
    id: str
    name: str

    def __init__(self, id, name):
        self.id = id
        self.name = name


class OecdDataResponseStructureAnnotation:
    title: str
    uri: str
    text: str

    def __init__(self, title, uri, text):
        self.title = title
        self.uri = uri
        self.text = text


class OecdDataResponseStructureAttributeObservation:
    id: str
    name: str
    values: List[OecdDataResponseValue]

    def __init__(self, id, name, values):
        self.id = id
        self.name = name
        self.values = values


class OecdDataResponseStructureAttributeSeries:
    id: str
    name: str
    values: List[OecdDataResponseValue]
    role: str
    default: str

    def __init__(self, id, name, values, role = None, default = None):
        self.id = id
        self.name = name
        self.values = values
        self.role = role
        self.default = default


class OecdDataResponseStructureAttributeDataSet:
    id: str
    name: str

    def __init__(self, id, name):
        self.id = id
        self.name = name


class OecdDataResponseStructureAttribute:
    dataSet: List[OecdDataResponseStructureAttributeDataSet]
    series: List[OecdDataResponseStructureAttributeSeries]
    observation: List[OecdDataResponseStructureAttributeObservation]

    def __init__(self, dataSet, series, observation):
        self.dataSet = dataSet
        self.series = series
        self.observation = observation


class OecdDataResponseStructureDimensionObservation:
    id: str
    name: str
    values: List[OecdDataResponseValue]
    role: str

    def __init__(self, id, name, values, role):
        self.id = id
        self.name = name
        self.values = values
        self.role = role


class OecdDataResponseStructureDimensionSeries:
    keyPosition: int
    id: str
    name: str
    values: List[OecdDataResponseValue]
    role: str

    def __init__(self, keyPosition, id, name, values, role = None):
        self.keyPosition = keyPosition
        self.id = id
        self.name = name
        self.values = values
        self.role = role


class OecdDataResponseStructureDimension:
    series: List[OecdDataResponseStructureDimensionSeries]
    observation: List[OecdDataResponseStructureDimensionObservation]

    def __init__(self, series, observation):
        self.series = series
        self.observation = observation


class OecdDataResponseDataSetSeriesValue:
    attributes: List[Any] # List[int or None]
    observations: Dict[int, List[Any]] # List[float or None]

    def __init__(self, attributes, observations):
        self.attributes = attributes
        self.observations = observations


class OecdDataResponseHeaderSender:
    id: str
    name: str

    def __init__(self, id, name):
        self.id = id
        self.name = name


class OecdDataResponseLink:
    href: str
    rel: str

    def __init__(self, href, rel):
        self.href = href
        self.rel = rel


class OecdDataResponseStructure:
    links: List[OecdDataResponseLink]
    name: str
    description: str
    dimensions: OecdDataResponseStructureDimension
    attributes: OecdDataResponseStructureAttribute
    annotations: List[OecdDataResponseStructureAnnotation]

    def __init__(self, links, name, description, dimensions, attributes, annotations):
        self.links = links
        self.name = name
        self.description = description
        self.dimensions = dimensions
        self.attributes = attributes
        self.annotations = annotations


# https://stackoverflow.com/questions/53905951/how-do-i-serialise-a-nested-dictionary-with-marshmallow
class OecdDataResponseDataSet:
    action: str
    series: Dict[str, OecdDataResponseDataSetSeriesValue]

    def __init__(self, action, series):
        self.action = action
        self.series = series


class OecdDataResponseHeader:
    id: str # uuid
    test: bool
    prepared: datetime
    sender: OecdDataResponseHeaderSender
    links: List[OecdDataResponseLink]

    def __init__(self, id, test, prepared, sender, links):
        self.id = id
        self.test = test
        self.prepared = prepared
        self.sender = sender
        self.links = links


class OecdDataResponse:
    header: OecdDataResponseHeader
    dataSets: List[OecdDataResponseDataSet]
    structure: OecdDataResponseStructure

    def __init__(self, header, dataSets, structure):
        self.header = header
        self.dataSets = dataSets
        self.structure = structure

File: csharp/OECD-ETL/Oecd.Data.External/Dto/OecdDataResponse.cs
namespace Oecd.Data.External.Dto;
public class OecdDataResponse
{
    public OecdDataResponseHeader Header { get; set; } = null!;
    public List<OecdDataResponseDataSet> DataSets { get; set; } = null!;
    public OecdDataResponseStructure Structure { get; set; } = null!;
}

public class OecdDataResponseHeader
{
    public Guid Id { get; set; }
    public bool Test { get; set; }
    public DateTime Prepared { get; set; }
    public OecdDataResponseHeaderSender Sender { get; set; } = null!;
    public List<OecdDataResponseLink> Links { get; set; } = new List<OecdDataResponseLink>();
}

public class OecdDataResponseValue
{
    public string Id { get; set; } = null!;
    public string Name { get; set; } = null!;
}

public class OecdDataResponseDataSet
{
    public string Action { get; set; } = null!;
    public Dictionary<string, OecdDataResponseDataSetSeriesValue> Series { get; set; } = new Dictionary<string, OecdDataResponseDataSetSeriesValue>();
}

public class OecdDataResponseStructure
{
    public List<OecdDataResponseLink> Links { get; set; } = new List<OecdDataResponseLink>();
    public string Name { get; set; } = null!;
    public string Description { get; set; } = null!;
    public OecdDataResponseStructureDimension Dimensions { get; set; } = null!;
    public OecdDataResponseStructureAttribute Attributes { get; set; } = null!;
    public List<OecdDataResponseStructureAnnotation> Annotations { get; set; } = new List<OecdDataResponseStructureAnnotation>();
}

public class OecdDataResponseLink
{
    public string Href { get; set; } = null!;
    public string Rel { get; set; } = null!;
}

public class OecdDataResponseHeaderSender
{
    public string Id { get; set; } = null!;
    public string Name { get; set; } = null!;
}

public class OecdDataResponseDataSetSeriesValue
{
    public List<int?> Attributes { get; set; } = new List<int?>();
    public Dictionary<int, List<double?>> Observations { get; set; } = new Dictionary<int, List<double?>>();
}

public class OecdDataResponseStructureDimension
{
    public List<OecdDataResponseStructureDimensionSeries> Series { get; set; } = new List<OecdDataResponseStructureDimensionSeries>();
    public List<OecdDataResponseStructureDimensionObservation> Observation { get; set; } = new List<OecdDataResponseStructureDimensionObservation>();
}

public class OecdDataResponseStructureDimensionSeries
{
    public int KeyPosition { get; set; }
    public string Id { get; set; } = null!;
    public string Name { get; set; } = null!;
    public List<OecdDataResponseValue> Values { get; set; } = new List<OecdDataResponseValue>();
    public string Role { get; set; } = null!;
}

public class OecdDataResponseStructureDimensionObservation
{
    public string Id { get; set; } = null!;
    public string Name { get; set; } = null!;
    public List<OecdDataResponseValue> Values { get; set; } = new List<OecdDataResponseValue>();
    public string Role { get; set; } = null!;
}

public class OecdDataResponseStructureAttribute
{
    // "dataSet": []
    public List<OecdDataResponseStructureAttributeSeries> Series { get; set; } = new List<OecdDataResponseStructureAttributeSeries>();
    public List<OecdDataResponseStructureAttributeObservation> Observation { get; set; } = new List<OecdDataResponseStructureAttributeObservation>();
}

public class OecdDataResponseStructureAttributeSeries
{
    public string Id { get; set; } = null!;
    public string Name { get; set; } = null!;
    public List<OecdDataResponseValue> Values { get; set; } = new List<OecdDataResponseValue>();
}

public class OecdDataResponseStructureAttributeObservation
{
    public string Id { get; set; } = null!;
    public string Name { get; set; } = null!;
    public List<OecdDataResponseValue> Values { get; set; } = new List<OecdDataResponseValue>();
}

public class OecdDataResponseStructureAnnotation
{
    public string Title { get; set; } = null!;
    public string Uri { get; set; } = null!;
    public string Text { get; set; } = null!;
}
